include/swift/SILOptimizer/Analysis/Reachability.h

```diff
@@ -887,6 +887,35 @@ void IterativeBackwardReachability<Effects>::Result::setEffectForBlock(
   }
 }
 
+//===----------------------------------------------------------------------===//
+// MARK: findBarriersBeforeInstructions
+//===----------------------------------------------------------------------===//
+
+using llvm::ArrayRef;
+using llvm::function_ref;
+
+struct ReachableBarriers final {
+  /// Instructions which are barriers.
+  llvm::SmallVector<SILInstruction *, 4> instructions;
+
+  /// Blocks one of whose phis is a barrier.
+  llvm::SmallVector<SILBasicBlock *, 4> phis;
+
+  /// Boundary edges; edges such that
+  /// (1) the target block is reachable-at-begin
+  /// (2) at least one adjacent edge's target is not reachable-at-begin.
+  llvm::SmallVector<SILBasicBlock *, 4> edges;
+
+  ReachableBarriers() {}
+  ReachableBarriers(ReachableBarriers const &) = delete;
+  ReachableBarriers &operator=(ReachableBarriers const &) = delete;
+};
+
+void findBarriersBeforeInstructions(
```

Comment:
> Does the defBlock need to dominate, or is it an optional non-barrier stopping point? Presumably this could be extended to multiple defBlocks, and maybe we should point that out somewhere

Took this opportunity to extend `IterativeBackwardReachability` to handle a list of `initialBlocks`.  

Renamed the function to `findBarriersBackward`.  It now takes a list of `initialBlocks` too.  And added a bit of documentation to the function declaration.